File: packages/cosmo-pd101/src/lib/synth/modulePresets.ts
import type { LfoWaveform } from "@/lib/synth/bindings/synth";

export type ModulePresetModule =
	| "chorus"
	| "delay"
	| "reverb"
	| "phaser"
	| "vibrato"
	| "phaseMod"
	| "lfo1"
	| "lfo2"
	| "modEnv";

export type ModulePresetPatch = Record<string, unknown>;

export type ModulePresetDefinition<TPatch extends ModulePresetPatch> = {
	id: string;
	label: string;
	patch: TPatch;
};

export const CHORUS_PRESETS: ModulePresetDefinition<{
	chorus: { enabled: boolean; rate: number; depth: number; mix: number };
}>[] = [
	{
		id: "classicWide",
		label: "Classic Wide",
		patch: {
			chorus: { enabled: true, rate: 0.9, depth: 1.2, mix: 0.38 },
		},
	},
	{
		id: "slowShimmer",
		label: "Slow Shimmer",
		patch: {
			chorus: { enabled: true, rate: 0.35, depth: 2.1, mix: 0.44 },
		},
	},
	{
		id: "ensembleThick",
		label: "Ensemble Thick",
		patch: {
			chorus: { enabled: true, rate: 1.8, depth: 2.6, mix: 0.56 },
		},
	},
];

export const DELAY_PRESETS: ModulePresetDefinition<{
	delay: {
		enabled: boolean;
		time: number;
		feedback: number;
		mix: number;
		tapeMode: boolean;
		warmth: number;
	};
}>[] = [
	{
		id: "digitalSlap",
		label: "Digital Slap",
		patch: {
			delay: {
				enabled: true,
				time: 0.11,
				feedback: 0.22,
				mix: 0.27,
				tapeMode: false,
				warmth: 0.2,
			},
		},
	},
	{
		id: "tapeEcho",
		label: "Tape Echo",
		patch: {
			delay: {
				enabled: true,
				time: 0.34,
				feedback: 0.46,
				mix: 0.35,
				tapeMode: true,
				warmth: 0.72,
			},
		},
	},
	{
		id: "dubFeedback",
		label: "Dub Feedback",
		patch: {
			delay: {
				enabled: true,
				time: 0.52,
				feedback: 0.68,
				mix: 0.4,
				tapeMode: true,
				warmth: 0.55,
			},
		},
	},
];

export const REVERB_PRESETS: ModulePresetDefinition<{
	reverb: {
		enabled: boolean;
		mix: number;
		space: number;
		predelay: number;
		distance: number;
		character: number;
	};
}>[] = [
	{
		id: "smallRoom",
		label: "Small Room",
		patch: {
			reverb: {
				enabled: true,
				mix: 0.22,
				space: 0.32,
				predelay: 0.006,
				distance: 0.28,
				character: 0.45,
			},
		},
	},
	{
		id: "plateAir",
		label: "Plate Air",
		patch: {
			reverb: {
				enabled: true,
				mix: 0.31,
				space: 0.58,
				predelay: 0.012,
				distance: 0.4,
				character: 0.74,
			},
		},
	},
	{
		id: "cathedral",
		label: "Cathedral",
		patch: {
			reverb: {
				enabled: true,
				mix: 0.47,
				space: 0.9,
				predelay: 0.03,
				distance: 0.68,
				character: 0.66,
			},
		},
	},
];

export const PHASER_PRESETS: ModulePresetDefinition<{
	phaser: {
		enabled: boolean;
		rate: number;
		depth: number;
		feedback: number;
		mix: number;
	};
}>[] = [
	{
		id: "gentleSweep",
		label: "Gentle Sweep",
		patch: {
			phaser: {
				enabled: true,
				rate: 0.35,
				depth: 0.45,
				feedback: 0.2,
				mix: 0.25,
			},
		},
	},
	{
		id: "jetWash",
		label: "Jet Wash",
		patch: {
			phaser: {
				enabled: true,
				rate: 0.9,
				depth: 0.78,
				feedback: 0.55,
				mix: 0.43,
			},
		},
	},
	{
		id: "wideNotch",
		label: "Wide Notch",
		patch: {
			phaser: {
				enabled: true,
				rate: 0.18,
				depth: 1,
				feedback: 0.72,
				mix: 0.52,
			},
		},
	},
];

export const VIBRATO_PRESETS: ModulePresetDefinition<{
	vibrato: {
		enabled: boolean;
		waveform: number;
		rate: number;
		depth: number;
		delay: number;
	};
}>[] = [
	{
		id: "subtle",
		label: "Subtle",
		patch: {
			vibrato: { enabled: true, waveform: 1, rate: 20, depth: 6, delay: 160 },
		},
	},
	{
		id: "chorused",
		label: "Chorused",
		patch: {
			vibrato: { enabled: true, waveform: 2, rate: 38, depth: 14, delay: 80 },
		},
	},
	{
		id: "warble",
		label: "Warble",
		patch: {
			vibrato: { enabled: true, waveform: 4, rate: 62, depth: 26, delay: 20 },
		},
	},
];

export const PHASE_MOD_PRESETS: ModulePresetDefinition<{
	intPmEnabled: boolean;
	intPmAmount: number;
	intPmRatio: number;
	pmPre: boolean;
}>[] = [
	{
		id: "glassBell",
		label: "Glass Bell",
		patch: {
			intPmEnabled: true,
			intPmAmount: 0.06,
			intPmRatio: 2.0,
			pmPre: true,
		},
	},
	{
		id: "metalFold",
		label: "Metal Fold",
		patch: {
			intPmEnabled: true,
			intPmAmount: 0.11,
			intPmRatio: 2.7,
			pmPre: true,
		},
	},
	{
		id: "aggressiveSync",
		label: "Aggressive Sync",
		patch: {
			intPmEnabled: true,
			intPmAmount: 0.18,
			intPmRatio: 3.4,
			pmPre: false,
		},
	},
];

export const LFO_PRESETS: ModulePresetDefinition<{
	waveform: LfoWaveform;
	rate: number;
	depth: number;
	symmetry: number;
	retrigger: boolean;
	offset: number;
}>[] = [
	{
		id: "slowSine",
		label: "Slow Sine",
		patch: {
			waveform: "sine",
			rate: 0.6,
			depth: 0.23,
			symmetry: 0.5,
			retrigger: false,
			offset: 0,
		},
	},
	{
		id: "tempoTri",
		label: "Tempo Tri",
		patch: {
			waveform: "triangle",
			rate: 2.25,
			depth: 0.48,
			symmetry: 0.5,
			retrigger: true,
			offset: 0,
		},
	},
	{
		id: "randomDrift",
		label: "Random Drift",
		patch: {
			waveform: "random",
			rate: 4.4,
			depth: 0.36,
			symmetry: 0.5,
			retrigger: false,
			offset: 0,
		},
	},
];

export const MOD_ENV_PRESETS: ModulePresetDefinition<{
	modEnv: { attack: number; decay: number; sustain: number; release: number };
}>[] = [
	{
		id: "pluck",
		label: "Pluck",
		patch: {
			modEnv: { attack: 0.005, decay: 0.16, sustain: 0.08, release: 0.14 },
		},
	},
	{
		id: "pad",
		label: "Pad",
		patch: {
			modEnv: { attack: 0.7, decay: 1.2, sustain: 0.75, release: 1.5 },
		},
	},
	{
		id: "reverseSwell",
		label: "Reverse Swell",
		patch: {
			modEnv: { attack: 1.8, decay: 0.28, sustain: 0.66, release: 0.95 },
		},
	},
];

export function getLfoModulePatch(id: 1 | 2, patch: Record<string, unknown>) {
	return id === 1 ? { lfo: patch } : { lfo2: patch };
}

File: packages/cosmo-synth-engine/src/module_presets.rs
use crate::params::{LfoWaveform, SynthParams};

pub fn apply_module_preset(params: &mut SynthParams, module: &str, preset: &str) -> bool {
    match module {
        "chorus" => apply_chorus_preset(params, preset),
        "delay" => apply_delay_preset(params, preset),
        "reverb" => apply_reverb_preset(params, preset),
        "phaser" => apply_phaser_preset(params, preset),
        "vibrato" => apply_vibrato_preset(params, preset),
        "phaseMod" => apply_phase_mod_preset(params, preset),
        "lfo1" => apply_lfo_preset(params, preset, false),
        "lfo2" => apply_lfo_preset(params, preset, true),
        "modEnv" => apply_mod_env_preset(params, preset),
        _ => false,
    }
}

fn apply_chorus_preset(params: &mut SynthParams, preset: &str) -> bool {
    match preset {
        "classicWide" => {
            params.chorus.enabled = true;
            params.chorus.rate = 0.9;
            params.chorus.depth = 1.2;
            params.chorus.mix = 0.38;
            true
        }
        "slowShimmer" => {
            params.chorus.enabled = true;
            params.chorus.rate = 0.35;
            params.chorus.depth = 2.1;
            params.chorus.mix = 0.44;
            true
        }
        "ensembleThick" => {
            params.chorus.enabled = true;
            params.chorus.rate = 1.8;
            params.chorus.depth = 2.6;
            params.chorus.mix = 0.56;
            true
        }
        _ => false,
    }
}

fn apply_delay_preset(params: &mut SynthParams, preset: &str) -> bool {
    match preset {
        "digitalSlap" => {
            params.delay.enabled = true;
            params.delay.time = 0.11;
            params.delay.feedback = 0.22;
            params.delay.mix = 0.27;
            params.delay.tape_mode = false;
            params.delay.warmth = 0.2;
            true
        }
        "tapeEcho" => {
            params.delay.enabled = true;
            params.delay.time = 0.34;
            params.delay.feedback = 0.46;
            params.delay.mix = 0.35;
            params.delay.tape_mode = true;
            params.delay.warmth = 0.72;
            true
        }
        "dubFeedback" => {
            params.delay.enabled = true;
            params.delay.time = 0.52;
            params.delay.feedback = 0.68;
            params.delay.mix = 0.4;
            params.delay.tape_mode = true;
            params.delay.warmth = 0.55;
            true
        }
        _ => false,
    }
}

fn apply_reverb_preset(params: &mut SynthParams, preset: &str) -> bool {
    match preset {
        "smallRoom" => {
            params.reverb.enabled = true;
            params.reverb.mix = 0.22;
            params.reverb.space = 0.32;
            params.reverb.predelay = 0.006;
            params.reverb.distance = 0.28;
            params.reverb.character = 0.45;
            true
        }
        "plateAir" => {
            params.reverb.enabled = true;
            params.reverb.mix = 0.31;
            params.reverb.space = 0.58;
            params.reverb.predelay = 0.012;
            params.reverb.distance = 0.4;
            params.reverb.character = 0.74;
            true
        }
        "cathedral" => {
            params.reverb.enabled = true;
            params.reverb.mix = 0.47;
            params.reverb.space = 0.9;
            params.reverb.predelay = 0.03;
            params.reverb.distance = 0.68;
            params.reverb.character = 0.66;
            true
        }
        _ => false,
    }
}

fn apply_phaser_preset(params: &mut SynthParams, preset: &str) -> bool {
    match preset {
        "gentleSweep" => {
            params.phaser.enabled = true;
            params.phaser.rate = 0.35;
            params.phaser.depth = 0.45;
            params.phaser.feedback = 0.2;
            params.phaser.mix = 0.25;
            true
        }
        "jetWash" => {
            params.phaser.enabled = true;
            params.phaser.rate = 0.9;
            params.phaser.depth = 0.78;
            params.phaser.feedback = 0.55;
            params.phaser.mix = 0.43;
            true
        }
        "wideNotch" => {
            params.phaser.enabled = true;
            params.phaser.rate = 0.18;
            params.phaser.depth = 1.0;
            params.phaser.feedback = 0.72;
            params.phaser.mix = 0.52;
            true
        }
        _ => false,
    }
}

fn apply_vibrato_preset(params: &mut SynthParams, preset: &str) -> bool {
    match preset {
        "subtle" => {
            params.vibrato.enabled = true;
            params.vibrato.waveform = 1;
            params.vibrato.rate = 20.0;
            params.vibrato.depth = 6.0;
            params.vibrato.delay = 160.0;
            true
        }
        "chorused" => {
            params.vibrato.enabled = true;
            params.vibrato.waveform = 2;
            params.vibrato.rate = 38.0;
            params.vibrato.depth = 14.0;
            params.vibrato.delay = 80.0;
            true
        }
        "warble" => {
            params.vibrato.enabled = true;
            params.vibrato.waveform = 4;
            params.vibrato.rate = 62.0;
            params.vibrato.depth = 26.0;
            params.vibrato.delay = 20.0;
            true
        }
        _ => false,
    }
}

fn apply_phase_mod_preset(params: &mut SynthParams, preset: &str) -> bool {
    match preset {
        "glassBell" => {
            params.int_pm_enabled = true;
            params.pm_pre = true;
            params.int_pm_amount = 0.06;
            params.int_pm_ratio = 2.0;
            true
        }
        "metalFold" => {
            params.int_pm_enabled = true;
            params.pm_pre = true;
            params.int_pm_amount = 0.11;
            params.int_pm_ratio = 2.7;
            true
        }
        "aggressiveSync" => {
            params.int_pm_enabled = true;
            params.pm_pre = false;
            params.int_pm_amount = 0.18;
            params.int_pm_ratio = 3.4;
            true
        }
        _ => false,
    }
}

fn apply_lfo_preset(params: &mut SynthParams, preset: &str, secondary: bool) -> bool {
    let lfo = if secondary {
        &mut params.lfo2
    } else {
        &mut params.lfo
    };

    match preset {
        "slowSine" => {
            lfo.waveform = LfoWaveform::Sine;
            lfo.rate = 0.6;
            lfo.depth = 0.23;
            lfo.symmetry = 0.5;
            lfo.retrigger = false;
            lfo.offset = 0.0;
            true
        }
        "tempoTri" => {
            lfo.waveform = LfoWaveform::Triangle;
            lfo.rate = 2.25;
            lfo.depth = 0.48;
            lfo.symmetry = 0.5;
            lfo.retrigger = true;
            lfo.offset = 0.0;
            true
        }
        "randomDrift" => {
            lfo.waveform = LfoWaveform::Random;
            lfo.rate = 4.4;
            lfo.depth = 0.36;
            lfo.symmetry = 0.5;
            lfo.retrigger = false;
            lfo.offset = 0.0;
            true
        }
        _ => false,
    }
}

fn apply_mod_env_preset(params: &mut SynthParams, preset: &str) -> bool {
    match preset {
        "pluck" => {
            params.mod_env.attack = 0.005;
            params.mod_env.decay = 0.16;
            params.mod_env.sustain = 0.08;
            params.mod_env.release = 0.14;
            true
        }
        "pad" => {
            params.mod_env.attack = 0.7;
            params.mod_env.decay = 1.2;
            params.mod_env.sustain = 0.75;
            params.mod_env.release = 1.5;
            true
        }
        "reverseSwell" => {
            params.mod_env.attack = 1.8;
            params.mod_env.decay = 0.28;
            params.mod_env.sustain = 0.66;
            params.mod_env.release = 0.95;
            true
        }
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn applies_known_presets() {
        let mut params = SynthParams::default();

        assert!(apply_module_preset(&mut params, "chorus", "classicWide"));
        assert!(params.chorus.enabled);

        assert!(apply_module_preset(&mut params, "delay", "tapeEcho"));
        assert!(params.delay.tape_mode);

        assert!(apply_module_preset(&mut params, "reverb", "cathedral"));
        assert!(params.reverb.mix > 0.4);

        assert!(apply_module_preset(&mut params, "phaser", "wideNotch"));
        assert!(params.phaser.depth > 0.9);

        assert!(apply_module_preset(&mut params, "vibrato", "warble"));
        assert_eq!(params.vibrato.waveform, 4);

        assert!(apply_module_preset(&mut params, "phaseMod", "metalFold"));
        assert!(params.int_pm_enabled);

        assert!(apply_module_preset(&mut params, "lfo1", "randomDrift"));
        assert_eq!(params.lfo.waveform, LfoWaveform::Random);

        assert!(apply_module_preset(&mut params, "lfo2", "tempoTri"));
        assert_eq!(params.lfo2.waveform, LfoWaveform::Triangle);

        assert!(apply_module_preset(&mut params, "modEnv", "pad"));
        assert!(params.mod_env.sustain > 0.7);
    }

    #[test]
    fn rejects_unknown_module_or_preset() {
        let mut params = SynthParams::default();
        assert!(!apply_module_preset(&mut params, "unknown", "x"));
        assert!(!apply_module_preset(&mut params, "chorus", "unknown"));
    }
}
